compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3689,7 +3689,7 @@ private boolean isAssignableToUnionType(BType source, BType target, Set<TypePair
                     }
                 }
                 // readonly can match to a union similar to any|error
-                if (sMember.tag == TypeTags.READONLY) {
+                if (sMember.tag == TypeTags.READONLY && isAssignable(symTable.anyOrErrorType, targetUnion)) {
```

Comment:
While this will work, I'm not sure if this is the best fix. The original issue is because true is returned incorrectly because there is an entry in unresolvedTypes, right? May be related to https://github.com/ballerina-platform/ballerina-lang/issues/36069 in a way too.

@KRVPerera, thoughts please? Since you're familiar with the union type checking logic also.

Reply:
Also, `any|error` isn't what is equivalent to `readonly`. It is `(any & readonly)|error`. Any specific reason to check with `symTable.anyOrErrorType`?

Reply:
Assigning to `(any & readonly)|error` is checked prior to this, this code snippet is only used for assigning `readonly` to `any|error` unions right?

Reply:
I don't think this check is correct. If `any|error` is a subtype of `targetUnion`, any type is a subtype of `targetUnion`. But, since the value space represented by `readonly` is a proper subset of that of `any|error`, just because `symTable.anyOrErrorType` is not a subtype of `targetUnion`, doesn't mean `readonly` isn't a subtype of `targetUnion`.

Why do we need this block anyway? What was the reason to check specifically with `symTable.anyOrErrorType`?

Reply:
Would it make sense to directly return `true` here if the `targetUnion` is `(any|error)` (regardless of the source type) ?

Reply:
That's a check we do in `isAssignable` anyway. I would have thought the check we need to do here is if `sMember.tag == TypeTags.READONLY` check if `(any & readonly)|error` is assignable to `targetUnion`.

@KRVPerera, would you happen to recall what we were trying to address here?

Reply:
We only check in `isAssignable` if `sourceType` doesn't contain an error type right? Without this check 
```
readonly x = 2;
any|error y = x;
```
will fail.

Reply:
Since readonly is also a complex union it may not match individual members of the target union but may be assignable to the whole target union as a whole. This is similar to the previous lines checking JSON.

Reply:
Didn't we discuss to use `(any & readonly)|error` instead of using `readonly` in that case? The newly added check wouldn't completely fix the original issue, right? Since we'll only be allowing `readonly` with target unions that are `any|error`?

Reply:
Yes, Ushira said the original problem is that code is adding readonly, target union pair. Yes, I remember but I think we had a discussion not for this code but for some change after this may be runtime. Need to fix the logic too.

Reply:
Yes looks like I have added it in runtime type checker. @ushirask Check runtime type checker too. There is `ANY_AND_READONLY_OR_ERROR_TYPE`

Reply:
I think chiran has adde `anyAndReadonlyOrError` to typeChecker too